src/libp2p/metrics.rs

```diff
@@ -33,3 +37,33 @@ pub static BAD_PEERS: Lazy<Gauge> = Lazy::new(|| {
     );
     metric
 });
+
+pub static PEER_TIPSET_EPOCH: Lazy<Family<PeerLabel, Gauge>> = Lazy::new(|| {
+    let metric = Family::default();
+    crate::metrics::default_registry().register(
+        "peer_tipset_epoch",
+        "peer tipset epoch",
+        metric.clone(),
+    );
+    metric
+});
+
+#[derive(Clone, Debug, Hash, PartialEq, Eq)]
+pub struct PeerLabel(PeerId);
+
+impl PeerLabel {
+    pub const fn new(peer: PeerId) -> Self {
+        Self(peer)
+    }
+}
+
+impl EncodeLabelSet for PeerLabel {
+    fn encode(&self, mut encoder: LabelSetEncoder) -> Result<(), std::fmt::Error> {
+        let mut label_encoder = encoder.encode_label();
+        let mut label_key_encoder = label_encoder.encode_label_key()?;
+        EncodeLabelKey::encode(&"PEER", &mut label_key_encoder)?;
+        let mut label_value_encoder = label_key_encoder.encode_label_value()?;
+        EncodeLabelValue::encode(&self.0.to_string(), &mut label_value_encoder)?;
```

Comment:
same as above, it's a bit confusing and there's totally a trait implementation for a variety of string flavours.

Reply:
Ah, I can see now that this is super ambiguous. A pity. 

Still, you don't need to pass this by reference, this should be enough as there is an impl for "String":
```
EncodeLabelValue::encode(self.0.to_string(), &mut label_value_encoder)?;
```

Reply:
Same, this does not compile

```
error[E0308]: mismatched types
   --> src/libp2p/metrics.rs:66:34
    |
66  |         EncodeLabelValue::encode(self.0.to_string(), &mut label_value_encoder)?;
    |         ------------------------ ^^^^^^^^^^^^^^^^^^ expected `&_`, found `String`
    |         |
    |         arguments to this function are incorrect
    |
    = note: expected reference `&_`
```

Reply:
Although the trait is implemented for `String`, the trait methods takes a borrowed self
fn [encode](https://docs.rs/prometheus-client/latest/prometheus_client/encoding/trait.EncodeLabelKey.html#tymethod.encode)(&self, encoder: &mut [LabelKeyEncoder](https://docs.rs/prometheus-client/latest/prometheus_client/encoding/struct.LabelKeyEncoder.html)<'_>) -> [Result](https://doc.rust-lang.org/nightly/core/result/enum.Result.html)<[()](https://doc.rust-lang.org/nightly/std/primitive.unit.html), [Error](https://doc.rust-lang.org/nightly/core/fmt/struct.Error.html)>;

Reply:
Yeah, what's the point of those implementations is unclear then.

Reply:
I can see now. This is pretty confusing looking at the code that's within those `encode` implementations. 
The way it's currently written is it wants to boil down to  `&str` impl, which is where it actually does ` encoder.write_str(self)?;`.